slides/supervised-regression/rsrc/poly_reg_example.R

```diff
@@ -0,0 +1,152 @@
+
+
+# Code for CV
+rmse <- function(observed, predicted) {
+  return(sqrt(sum((observed - predicted)^2) / length(observed)))
+}
+
+
+get_folds <- function(n, K = n_folds) {
+  cv_folds <- cvTools::cvFolds(n = n, K = K)
+  cv_folds <- split(x = cv_folds$subsets[, 1], f = cv_folds$which)
+  return(cv_folds)
+}
+
+
+do_cv <- function(data, folds, degree0) {
+  # Berechnung des Mittelwertes der RMSEs aller Teildatens¨atze:
+  cv_mod_test <- mean(vapply(X = folds, FUN = function(i) {
+    model <- lm(rentals ~ poly(temp, degree = degree0) , data = data[-i, ])
+    rmse(observed = data[i, "rentals"],
+         predicted = predict(object = model, newdata = data[i,]))
+  }, FUN.VALUE = numeric(1)))
+  
+  cv_mod_train <- mean(vapply(X = folds, FUN = function(i) {
+    model <- lm(rentals ~ poly(temp, degree = degree0) , data = data[-i, ])
+    rmse(observed = data[-i, "rentals"],
+         predicted = predict(object = model, newdata = data[-i,]))
+  }, FUN.VALUE = numeric(1)))
+  
+  # fit model on whole data set
+  mod_data = lm(rentals ~ poly(temp, degree = degree0) , data = data)
+  
+  # generate effect plots:
+  plot <- effect_plot(mod_data, pred = temp, data = data, plot.points = TRUE, 
+              interval = TRUE, line.color = "blue", y.label = "number of rentals",
+              x.lab  = "normalized temperature", main.title = paste0("effect plot | degree: ", degree0))
+  
+  
+  
+  # save coefficient estimates:
+  estimated_effects <- mod_data$coefficients
+  
+  
+  # Ausgabe des Kreuzvalidierungs-RMSEs:
+  attr(x = cv_mod_test, which = "degree") <- degree0
+  attr(x = cv_mod_train, which = "degree") <- degree0
+  attr(x = cv_mod_test, which = "data") <- "test set: predicted GE"
+  attr(x = cv_mod_train, which = "data") <- "train set: training error"
+  return(list(cv_mod_test, cv_mod_train, estimated_effects, plot))
+}
+
+
+
+do_cvs <- function(data, folds, degrees) {
+  # Generierung aller m¨oglichen Modellspezifikationen:
+  
+  # Durchf¨uhrung der Kreuzvalidierung f¨ur die einzelnen Modellspezifikationen:
+  res_all <- lapply(degrees, FUN = function(z) do_cv(data = data, folds = folds, degree0 = z))
+  # Extraktion des gem¨aß des RMSEs besten Modells:
+  # model_best <- lm(formula = attr(x = get_best(res_all), which = "formula"),
+  #                  data = data)
+  # return(coef(model_best))
+  return(res_all)
+  
+}
+
+
+# run CV + plots + coefficient estimates
+# set highest degree for polynomials
+degrees0 <- seq_len(8)
+# set number of folds for CV
+n_folds <- 10
+# run CV
+set.seed(1)
+folds0 = get_folds(n = nrow(bike_data))
+set.seed(1)
+CV_results <- do_cvs(bike_data, folds = folds0, degrees = degrees0)
+
+# extract results for train and test error
+error_list <- lapply(CV_results, function(i) i[c(1,2)])
+error_dt <- rbindlist(error_list)
+error_dt <- cbind(degrees0, error_dt)
+colnames(error_dt) <- c("degree", "test_error", "train_error")
+
+# plotting test and training error
+jpeg(file= "plot: train vs test error.jpeg")
```

Comment:
gern ggsave benutzen (s. nächster kommentar, ggplot is the way to go). außerdem in den ordner "figures" speichern und bitte keine filenames mit doppelpunkt, leerzeichen etc, am besten einfach underscores